base/sparse/higherorderfns.jl

```diff
@@ -888,40 +889,129 @@ broadcast{Tf,T}(f::Tf, ::Type{T}, A::SparseMatrixCSC) = broadcast(y -> f(T, y),
+# such methods whenever extending broadcast to a new type, which ultimately results
+# in having to write a combinatorial explosion of ambiguity-killing methods.
+#
+# Perhaps the following makes a reasonable model: don't define methods like
+# promote_containertype(ct, ::Type{Array}) = Array, and discourage definition of such
+# methods for new types. Instead (1) define promote_containertype(cta, ctb) = AbstractArray as
+# primary fallback, such that any type pair without clearly defined behavior gets funneled
+# to the generic AbstractArray broadcast code, and (2) encourage writing only explicit, tight
+# promote_containertype definitions.
+#
+# The above should improve the extensibility and maintainability of Broadcast.
+#
+# The implementation of extending sparse broadcast to handle Arrays below illustrates the
+# above issue. Specifically, here we need to be able to identify Arrays (in sense (1) above),
+# and specifically one- and two-dimensional Arrays (though sparse broadcast can handle the
+# dimension separation internally just fine), so we introduce _containertype methods for
+# Vector and Matrix. But to keep existing code functioning, this forces all other consumers
+# of broadcast that define methods like promote_containertype(ct, ::Type{Custom}) to be
+# aware of Vector and Matrix (see e.g. the modifications necessary to test/broadcast.jl)
+# in addition to their present awareness of Array.
+#
+# So here's the problematic hack, which impacts other consumers of Broadcast if they define
+# additional ambiguity-magnets like promote_containertype(ct, ::Type{Custom}). Such consumers
+# will need to define disambiguating methods for Vector and Matrix as they now do for Array.
+_containertype{T<:Vector}(::Type{T}) = Vector
+_containertype{T<:Matrix}(::Type{T}) = Matrix
+broadcast_indices(::Type{Vector}, A) = indices(A)
+broadcast_indices(::Type{Matrix}, A) = indices(A)
+@inline broadcast_c(f, ::Type{Vector}, A, Bs...) = broadcast_c(f, Array, A, Bs...)
+@inline broadcast_c(f, ::Type{Matrix}, A, Bs...) = broadcast_c(f, Array, A, Bs...)
+
+promote_containertype(::Type{Vector}, ct) = Array
+promote_containertype(ct, ::Type{Vector}) = Array
+promote_containertype(::Type{Vector}, ::Type{Vector}) = Vector
+promote_containertype(::Type{Vector}, ::Type{Array}) = Array
+promote_containertype(::Type{Array}, ::Type{Vector}) = Array
+promote_containertype(::Type{Vector}, ::Type{Any}) = Vector
+promote_containertype(::Type{Any}, ::Type{Vector}) = Vector
+
+promote_containertype(::Type{Matrix}, ct) = Array
+promote_containertype(ct, ::Type{Matrix}) = Array
+promote_containertype(::Type{Matrix}, ::Type{Matrix}) = Matrix
+promote_containertype(::Type{Matrix}, ::Type{Array}) = Array
+promote_containertype(::Type{Array}, ::Type{Matrix}) = Array
+promote_containertype(::Type{Matrix}, ::Type{Any}) =  Matrix
+promote_containertype(::Type{Any}, ::Type{Matrix}) = Matrix
+
+promote_containertype(::Type{Vector}, ::Type{Matrix}) = Matrix
```

Comment:
Couldn't you replace definitions like this one with a general definition along the lines of this? Would it help fixing the problem?
```julia
promote_containertype{S<:Array, T<:Array}(::Type{S}, ::Type{T}) =
    Array{promote_type(eltype(S), eltype(T)), max(ndim(S), ndim(T))}`
```

(The element type isn't needed here of course, but I don't know how to create the `Array{T, N}` type without speciying `T`.)